### ISSUE TITLE
How can I implement refresh token mechanism in an GRPC android client?

### ISSUE THREAD

**Issue by Raj-Varun** [grpc/grpc-java#6638]
First of all, there are already two questions on the StackOverflow with this same kinda issue [this](https://stackoverflow.com/q/50151275/7436566) and [this](https://stackoverflow.com/q/51840892/7436566) which is kinda workaround, not an actual solution.

So there is a `ClientInterceptor` that is intercepting and attaching the header to the request so if the `access token` expired the server will respond with an `UNAUTHENTICATED` status but then how to recall the same method with new `access token` attached to it.
It is something like an authenticator which is used in retrofit here is the  [link.](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-authenticator/)

**Comment by creamsoup** [comment]
you may find [this thread](https://github.com/grpc/grpc-java/issues/5856) useful.

**Comment by Raj-Varun** [comment]
Hey, @creamsoup thanks for your response it really means a lot. Actually, when I'm trying to convert the class from the answer to kotlin I'm getting this error. Can you help me with this?

`
Type inference failed. Expected type mismatch: inferred type is (ClientCall<(ReqT#2 (type parameter of com.app.android.app.services.Interceptor.interceptCall)..ReqT#2?), (RespT#2 (type parameter of com.app.android.app.services.Interceptor.interceptCall)..RespT#2?)>..ClientCall<ReqT#2!, RespT#2!>?) but ClientCall<ReqT#1 (type parameter of com.app.android.app.services.Interceptor.interceptCall.RetryingUnaryRequestClientCall), RespT#1 (type parameter of com.app.android.app.services.Interceptor.interceptCall.RetryingUnaryRequestClientCall)>? was expected
`

```
class Interceptor(private val app: App) : ClientInterceptor {

    override fun <ReqT : Any?, RespT : Any?> interceptCall(method: MethodDescriptor<ReqT, RespT>?, callOptions: CallOptions?, next: Channel?): ClientCall<ReqT, RespT> {

        class RetryingUnaryRequestClientCall<ReqT, RespT> : ClientCall<ReqT, RespT>() {

            var listener: Listener<RespT>? = null
            var metadata: Metadata? = null
            var message: ReqT? = null
            var request = 0
            var call: ClientCall<ReqT, RespT>? = null

            override fun sendMessage(message: ReqT) {
                assert(this.message == null)
                this.message = message
            }

            override fun halfClose() {
                startCall(CheckingListener())
            }

            override fun start(responseListener: Listener<RespT>?, headers: Metadata?) {
                this.listener = responseListener
                this.metadata = headers
            }

            override fun cancel(message: String?, cause: Throwable?) {
                if (call != null) {
                    call?.cancel(message, cause)
                }
                listener?.onClose(Status.CANCELLED.withDescription(message).withCause(cause), Metadata())
            }

            override fun request(numMessages: Int) {
                request += numMessages
                assert(this.message == null)
            }

            private fun startCall(listener: Listener<RespT>?) {
                call = next?.newCall(method,callOptions)  <---- This line
                val headers = Metadata()
                headers.merge(metadata)
                call?.start(listener, headers)
                assert(this.message != null)
                call?.request(request)
                call?.sendMessage(message)
                call?.halfClose()
            }

            inner class CheckingListener : ForwardingClientCallListener<RespT>() {

                var delegate: Listener<RespT>? = null

                override fun delegate(): Listener<RespT> {
                    if (delegate == null) {
                        throw IllegalStateException()
                    }
                    return delegate!!
                }

                override fun onHeaders(headers: Metadata?) {
                    delegate = listener
                    super.onHeaders(headers)
                }

                override fun onClose(status: Status?, trailers: Metadata?) {
                    if (delegate == null) {
                        super.onClose(status, trailers)
                        return
                    }
                    if (!needToRetry(status, trailers)) {
                        delegate = listener
                        super.onClose(status, trailers)
                        return
                    }
                    startCall(listener) // Only retry once
                }

                private fun needToRetry(status: Status?, trailers: Metadata?): Boolean {
                    if (status?.code?.toStatus() == UNAUTHENTICATED) {
                        Log.e("code", status?.code.toString())
                        return true
                    }
                    return false
                }
            }
        }
        return RetryingUnaryRequestClientCall()
    }
}
```

**Comment by creamsoup** [comment]
i am not familiar with kotlin. it looks like the last return statement is missing generic types (not sure kotlin fills it for you or not), do you need to specify types?

**Comment by Raj-Varun** [comment]
No, it doesn't require generic types!

**Comment by creamsoup** [comment]
yeah like I said i am not familiar with kotlin. the error message is related to the type inferring, is there a way to provide hint or solid type?

**Comment by Raj-Varun** [comment]
Yeah, I understand.
I think I know the issue but I don't know how to fix it.
If you look at the java code in the class `RetryingUnaryRequestClientCall` it has a member `ClientCall call;` but it kotlin if I do this `var call: ClientCall` (i.e declaring the member same as given in the java code) it gives this error 
```2 type arguments expected for class ClientCall<ReqT : Any!, RespT : Any!>``` but it fixes the old error.

**Comment by Raj-Varun** [comment]
Hey, @creamsoup thanks for your replies I have fixed it. I forgot about the best thing in kotlin which is using objects. So instead of using class I created the object of that class and passed it directly. Now it is showing no errors and this method is not calling up in kotlin
```
override fun onHeaders(headers: Metadata?) {
    delegate = listener
     super.onHeaders(headers)
}
```
So it was throwing the `IllegalStateException` so to assign the delegate I used the `onReady` method and now it looks fine!

Thanks again.
